main/guides/js-programming/hardened-js.md

```diff
@@ -0,0 +1,388 @@
+
+::: tip harden(), classes, and details
+Note that hardening a class instance also hardens the class.
+For more details, see [harden API in the `ses` package](https://github.com/endojs/endo/blob/master/packages/ses/README.md#harden)
+:::
+## Objects with state
+
+Now let's review the `makeCounter` example:
+
+<<< @/snippets/test-hardened-js.js#counterAnimation
+
+Each call to `makeCounter` creates a new encapsulated `count` variable
+along with `incr` and `decr` functions. The `incr` and `decr` functions [close over](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures) the `count` variable.
+
+To see how this works in detail, you may want to step through this
+[visualization of the code](https://pythontutor.com/live.html#code=const%20makeCounter%20%3D%20%28%29%20%3D%3E%20%7B%0A%20%20let%20count%20%3D%200%3B%0A%20%20const%20it%20%3D%20%7B%0A%20%20%20%20incr%3A%20%28%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20count%20%2B%3D%201%3B%0A%20%20%20%20%20%20return%20count%3B%0A%20%20%20%20%7D,%0A%20%20%20%20decr%3A%20%28%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20count%20-%3D%201%3B%0A%20%20%20%20%20%20return%20count%3B%0A%20%20%20%20%7D,%0A%20%20%7D%3B%0A%20%20return%20Object.freeze%28it%29%3B%0A%7D%3B%0A%0Aconst%20c1%20%3D%20makeCounter%28%29%3B%0A%0Aconst%20c2%20%3D%20makeCounter%28%29%3B%0Aconsole.log%28c2.incr%28%29%29%3B%0Aconsole.log%28%5Bc1.incr%28%29,%20c2.incr%28%29%5D%29%3B&cumulative=false&curInstr=25&heapPrimitives=nevernest&mode=display&origin=opt-live.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false"):
+
+[![makeCounter code animation](../../assets/counter-animation.png)](https://pythontutor.com/live.html#code=const%20makeCounter%20%3D%20%28%29%20%3D%3E%20%7B%0A%20%20let%20count%20%3D%200%3B%0A%20%20const%20it%20%3D%20%7B%0A%20%20%20%20incr%3A%20%28%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20count%20%2B%3D%201%3B%0A%20%20%20%20%20%20return%20count%3B%0A%20%20%20%20%7D,%0A%20%20%20%20decr%3A%20%28%29%20%3D%3E%20%7B%0A%20%20%20%20%20%20count%20-%3D%201%3B%0A%20%20%20%20%20%20return%20count%3B%0A%20%20%20%20%7D,%0A%20%20%7D%3B%0A%20%20return%20Object.freeze%28it%29%3B%0A%7D%3B%0A%0Aconst%20c1%20%3D%20makeCounter%28%29%3B%0A%0Aconst%20c2%20%3D%20makeCounter%28%29%3B%0Aconsole.log%28c2.incr%28%29%29%3B%0Aconsole.log%28%5Bc1.incr%28%29,%20c2.incr%28%29%5D%29%3B&cumulative=false&curInstr=25&heapPrimitives=nevernest&mode=display&origin=opt-live.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false")
+
+## Hardening JavaScript: strict mode
+
+The first step to hardening JavaScript is that Hardened JavaScript
+is always in [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode).
+
+[![Subsetting JavaScript](https://raw.githubusercontent.com/endojs/Jessie/main/docs/jessie.png)](https://github.com/endojs/Jessie#subsetting-ecmascript)
+
+The most likely way that you would notice this is if you
+make a typo in a variable name: this will throw a `ReferenceError`
+rather than silently continuing, with untold results.
```

Comment:
Sloppy mode does return a reference error for this too. Assigning to a variable, OTOH. OTGH, our repl is strict but for this assignment behavior, so this might not be the best example.

Failure to assign to a frozen property is best example, as it ties in to `harden` explanation.

Reply:
"with untold results" makes it sound like there's something unpredictable or non-deterministic going on.